bin/BuildPackages.sh

```diff
@@ -21,55 +50,82 @@ set -e
 # an idea what to do for a complete installation of GAP.
 
 
-if [ $# -eq 0 ]
-  then
-    echo "Assuming default GAP location: ../.."
-    GAPDIR=../..
-  else
-    echo "Using GAP location: $1"
-    GAPDIR="$1"
+# Is someone trying to run us from inside the 'bin' directory?
+if [[ -f gapicon.bmp ]]
+then
+  error "This script must be run from inside the pkg directory" \
+        "Type: cd ../pkg; ../bin/BuildPackages.sh"
 fi
 
-# Is someone trying to run us from inside the 'bin' directory?
-if [ -f gapicon.bmp ]
-  then
-    echo "This script must be run from inside the pkg directory"
-    echo "Type: cd ../pkg; ../bin/BuildPackages.sh"
-    exit 1
+# CURDIR
+CURDIR="$(pwd)"
+
+# GAPDIR
+if [[ "$#" -ge 1 ]]
+then
+  case "$1" in
+    --with-gaproot=*)
+      GAPDIR="$(sed 's|--with-gaproot=||' <<< $1)"
+      shift
+    ;;
+
+    *)
+      cd ..
+      GAPDIR="$(pwd)"
+      cd "$CURDIR"
+    ;;
+  esac
+else
+  cd ..
+  GAPDIR="$(pwd)"
+  cd "$CURDIR"
 fi
+notice "Using GAP location: $GAPDIR"
 
-# We need any subdirectory, to test if $GAPDIR is right
-SUBDIR=`ls -d */ | head -n 1`
-if ! (cd $SUBDIR && [ -f $GAPDIR/sysinfo.gap ])
-  then
-    echo "$GAPDIR is not the root of a gap installation (no sysinfo.gap)"
-    echo "Please provide the absolute path of your GAP root directory as"
-    echo "first argument to this script."
-    exit 1
+# pakcages to build
+if [[ "$#" -ge 1 ]]
+then
+  # last arguments are packages to build
+  PACKAGES=("$@")
+else
+  # if there are no arguments, then build all packages
+  shopt -s nullglob
+  PACKAGES=(*)
+  shopt -u nullglob
 fi
 
-if (cd $SUBDIR && grep 'ABI_CFLAGS=-m32' $GAPDIR/Makefile > /dev/null) ; then
-  echo "Building with 32-bit ABI"
+# We need to test if $GAPDIR is right
```

Comment:
Wouldn't it make more sense to do this right after setting GAPDIR, and before dealing with PACKAGES?